hw/syn/tools/dc/parse-syn-report.py

```diff
@@ -93,88 +93,145 @@ def _rel_err(val, ref):
-            ]
+    # current depth level of sub-modules
+    pattern = r"^(" + prefix + r"[\.0-9A-Za-z_\[\]]+){1}(?:(?:/[\.0-9A-Za-z_\[\]]+)*)"
 
     for k in range(5):
         pattern += r"\s+(" + FP_NUMBER + r")"
     matches = re.findall(pattern, full_file, flags=re.MULTILINE)
 
-    # checksums
-    comb_check = 0.0
-    reg_check = 0.0
-    macro_check = 0.0
+    # we drop the first entry as it always corresponds to the top-level
+    # and for that we already parsed out the summary numbers.
+    if matches and depth == 1:
+        matches = matches[1:]
+
     try:
-        for item in matches:
-            if item[0] not in results[key]["instances"]:
+        for inst in matches:
+            if inst[0] not in results[key]["instances"]:
                 results[key]["instances"].update({
-                    item[0]: {
+                    inst[0]: {
                         "comb": 0.0,
                         "reg": 0.0,
                         "buf": float("nan"),  # not available here
+                        "logic": 0.0,
                         "macro": 0.0,
                         "total": 0.0,
+                        "depth": depth
                     }
                 })
-            results[key]["instances"][item[0]]["comb"] += float(item[3])
-            results[key]["instances"][item[0]]["reg"] += float(item[4])
-            results[key]["instances"][item[0]]["macro"] += float(item[5])
-            results[key]["instances"][item[0]]["total"] += float(item[3]) + \
-                                                           float(item[4]) + \
-                                                           float(item[5])
-            comb_check += float(item[3])
-            reg_check += float(item[4])
-            macro_check += float(item[5])
+
+                # if we're not yet at depth, step one level down
+                # if this module has been specified
+                if inst[0] in args.expand_modules or depth < args.expand_depth:
+                    _extract_area_recursive(full_file,
+                                            results,
+                                            key,
+                                            args,
+                                            depth=depth + 1,
+                                            prefix=inst[0] + "/")
+
+            results[key]["instances"][inst[0]]["comb"] += float(inst[3])
+            results[key]["instances"][inst[0]]["reg"] += float(inst[4])
+            results[key]["instances"][inst[0]]["logic"] += float(inst[3]) + \
+                                                           float(inst[4])
+            results[key]["instances"][inst[0]]["macro"] += float(inst[5])
+            results[key]["instances"][inst[0]]["total"] += float(inst[3]) + \
+                                                           float(inst[4]) + \
+                                                           float(inst[5])
```

Comment:
All the float casts and such are getting a little out of hand here, and the continuation line indent doesn't match PEP8 (you'll see some flake8 errors here).

How about something like this?
```python3
    instances = results[key]['instances']
    try:
        for match in matches:
            name = match[0]

            if name not in instances:
                instances[name] = {
                    "comb": 0.0,
                    "reg": 0.0,
                    "buf": float("nan"),  # not available here
                    "logic": 0.0,
                    "macro": 0.0,
                    "total": 0.0,
                    "depth": depth
                }

                # if we're not yet at depth, step one level down
                # if this module has been specified
                if name in args.expand_modules or depth < args.expand_depth:
                    _extract_area_recursive(full_file,
                                            results,
                                            key,
                                            args,
                                            depth=depth + 1,
                                            prefix=name + "/")

            comb = float(match[3])
            reg = float(match[4])
            macro = float(match[5])

            instance = instances[name]

            instance["comb"] += comb
            instance["reg"] += reg
            instance["logic"] += comb + reg
            instance["macro"] += macro
            instance["total"] += comb + reg + macro

    except ValueError as err:
        results["messages"]["flow_errors"] += ["ValueError: %s" % err]
```

Reply:
yeah that's a bit clunky, I'll clean this up.